src/deephaven/plugin/object.py

```diff
@@ -17,46 +17,84 @@ def index(self) -> int:
+    likewise be given an instance of this interface to be able to send messages to the client.
+    """
+    def __init__(self):
+        pass
 
     @abc.abstractmethod
-    def reference(
-        self, object, allow_unknown_type: bool = False, force_new: bool = False
-    ) -> Optional[Reference]:
-        """Gets the reference for object if it has already been created and force_new is False,
-        otherwise creates a new one. If allow_unknown_type is False, and no type can be found, no
-        reference will be created."""
+    def on_close(self):
+        """Closes the stream on both ends. No further messages can be sent or received."""
+        pass
+
+    @abc.abstractmethod
+    def on_data(self, payload: bytes, references: List[Any]):
+        """Transmits data to the remote end of the stream. This can consist of a binary payload and references to
+        objects on the server.
+        """
         pass
 
 
 class ObjectType(Plugin):
     """An object type plugin. Useful for serializing custom objects between the server / client."""
 
+    def __init__(self):
+        pass
+
     @property
     @abc.abstractmethod
     def name(self) -> str:
         """The name of the object type."""
         pass
 
     @abc.abstractmethod
-    def is_type(self, object) -> bool:
-        """Returns true if, and only if, the object is compatible with this object type."""
+    def is_type(self, obj: Any) -> bool:
+        """Returns True if, and only if, the object is compatible with this object type."""
         pass
 
+
+class BidirectionObjectType(ObjectType):
+    """Base class for an object type that can continue to send responses to the client, or receive requests
+    from the server even after it is fetched.
+    """
+    def __init__(self):
+        pass
+    @abc.abstractmethod
+    def create_client_connection(self, obj: object, connection: MessageStream) -> MessageStream:
+        """Signals creation of a client stream to the specified object. The returned MessageStream implementation will
+        be called with each received message from the client, and can call the provided connection parameter to send
+        messages as needed to the client.
+
+        Before returning, this method must call connection.on_message with some initial payload, so that the client has
+        an initial view of the object.
+        """
+        pass
+
+
+class FetchOnlyObjectType(ObjectType):
+    """Base class for an object type which will only be fetched once, rather than support streaming requests or
+    responses.
+    """
     @abc.abstractmethod
-    def to_bytes(self, exporter: Exporter, object) -> bytes:
+    def to_bytes(self, exporter: Exporter, obj: Any) -> bytes:
         """Serializes object into bytes. Must only be called with a compatible object."""
         pass
 
 
-def find_object_type(object) -> Optional[ObjectType]:
+def find_object_type(obj: Any) -> Optional[ObjectType]:
```

Comment:
pydoc

Reply:
@devinrsmith what does this do? I can't find any invocation of it, is it meant to be user facing, or is it perhaps unused and okay to be removed?

Reply:
This is an end-user utility to lookup the ObjectType from any object:

```python
import pandas as pd
from deephaven.plugin.object import find_object_type

df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
print(find_object_type(df))
```

outputs

```
11:07:27.898 STDOUT <deephaven.pandasplugin.pandas_as_table.PandasDataFrameSerializer object at 0x7ff7a48d7e10>
```

At a minimum, I think it's a generally useful debugging tool.

It _is_ also used by deephaven-plugin-json testing to mock an Exporter, https://github.com/deephaven/deephaven-plugin-json/blob/v0.0.1.dev4/test/test_json_plugin.py#L33, although this may be unnecessary in the future b/c of the changes to the plugin interfaces.

Reply:
Unless I'm mistaken, I believe it is incomplete, since it only finds python plugins, and can't see server plugins? Shouldn't we provide this as a feature from the actual server?

Reply:
Yes, good point.